contents/projectManagement/testDrivenDevelopment.md

```diff
@@ -0,0 +1,169 @@
+Mock objects are simulated objects whose sole purpose is to act as a black box that mimics the behaviour of real-world objects during testing. They are especially useful in writing unit tests, which are meant to test the functionality of a portion of production code assuming that the rest of the code behaves as it should. 
+</modal>
+
+### Benefit 2: Focuses on Understanding Product Requirements
+
+Test-driven development requires the developers to have a good understanding of the product or feature requirements before any sort of development can begin. As mentioned earlier, the first step in development is writing tests. Coming up with tests without a clear picture of the requirements or specifications  can be dangerous as: 
+
+1. you could end up testing something that doesn't need to be tested, or 
+1. you could end up overlooking some requirements that _must_ be tested, thus giving you a false sense of accomplishment when you write production code to pass the test
+
+TDD thus helps in preventing teams from rushing into development and pushes them to prioritize the requirements-gathering step. 
+
+### Benefit 3: Increases Productivity
+
+A study done by researchers in CMU found that programmers who wrote more tests tended to be more productive^[Erdogmus, 2005. ["On the Effectiveness of Test-first Approach to Programming"](https://web.archive.org/web/20141222180731/http://nparc.cisti-icist.nrc-cnrc.gc.ca/npsi/ctrl?action=shwart&index=an&req=5763742&lang=en)], and since employing TDD meant writing more tests, there is a correlation between productivity and TDD.
+
+Test-driven development makes programmers think about the desired outcome of certain processes, features, and how other code can be integrated. Programmers will then think about coming up with failing tests and working their way up to writing code that passes the test. Employing test-driven development thus helps to get a better understanding of the design of a program [^3]. 
+
+As mentioned before, test-driven development can potentially result in software with high code coverage. This, in turn, can boost individual and team morale, confidence in the code, and productivity as a consequence.
+
+## Limitations
+
+Even with some of the attractive benefits of adopting TDD, there is no one-size-fits-all solution for project design management. Some of the limitations of TDD are: 
+
+### Limitation 1: Strict Emphasis on Communication of Requirements and Specifications
+
+Product/Feature requirements are crucial to teams that employ a TDD-style approach to development. As mentioned earlier, the very first step of the development cycle is to write acceptance tests. If the requirements are not communicated properly, teams will end up constantly having to change their test suites. In extreme scenarios, a re-write may be necessary.
+
+As a result, a lot of man hours can go into making sure the requirements are in order -  consequently, this could result in a slow start to development.
```

Comment:
Not convinced this is a limitation. Also, this is not true. TDD doesn't mean slow start to a project. TDD comes from the Agile World (it was invented by Kent Beck who also invented XP) which doesn't believe in heavy upfront requirements gathering.

Reply:
Hmm i guess it _does_ conflict with the `Benefit 2: Focuses on Understanding Product Requirements`. I will remove this, in that case.

Reply:
TDD doesn't mean you do the whole requirement thing in one go at the start. It forces you to understand the requirement of whatever small thing you implement, whether it is a user story, a method, class, etc.

Reply:
Right. Then I guess we should stick to how that can be a good thing (as in the 2nd benefit), and not portray it in a negative sense?